### DISCUSSION TITLE
ARDUINO-STK output to PWM or ESP32 DAC?

### DISCUSSION THREAD

**Discussion by SnazzyBlanket** [pschatzmann/arduino-stk#13] [General]
looking here 

https://github.com/pschatzmann/arduino-stk/wiki/Output-of-Sound

I only see text and I2s as an option. I would like to know how to output to the built in Dac on esp32 or a pwm pin 

thanks

**Comment by pschatzmann** [comment]
I would use the AudioTools for doing the Output:
- AnalogAudioStream is using the internal DAC 
- I2SStream is using i2s
- PWMAudioOutput is for pwm

**Reply by SnazzyBlanket** [reply]
thanks a million. I will install that now.

**Comment by SnazzyBlanket** [comment]
`#include <Arduino.h>


#include "WiFi.h"


/**
 * @file streams-generator_fromarray-analog.ino
 * @author Phil Schatzmann
 * @brief Example for GeneratorFromArray
 * @version 0.1
 * @date 2022-03-22
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"

AudioInfo info(44100, 1, 16);
int16_t sine_array[] = {0, 4560, 9031, 13327, 17363, 21062, 24350, 27165, 29450, 31163, 32269, 32747, 32587, 31793, 30381, 28377, 25820, 22761, 19259, 15383, 11206, 6812, 2285, -2285, -6812, -11206, -15383, -19259, -22761, -25820, -28377, -30381, -31793, -32587, -32747, -32269, -31163, -29450, -27165, -24350, -21062, -17363, -13327, -9031, -4560  };
GeneratorFromArray<int16_t> sineWave(sine_array,0,false);
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
AnalogAudioStream out; 
StreamCopy copier(out, sound);                             // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.copyFrom(info); 
  out.begin(config);

  // Setup sine wave
  sineWave.begin(info);
  Serial.println("started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}
`

**Comment by SnazzyBlanket** [comment]
I thought I would post that here for anyone who wants to get Audio working on a esp32 feather with Platform IO!

**Reply by SnazzyBlanket** [reply]
Now I am trying to get an STK example working. I noticed in the examples of the arduino audio tools there are STK examples. however they no longer seem to use TICK. 

and they include #include "AudioLibs/AudioSTK.h" instead of using namespace stk;

so I am not sure which one is correct? possibly STK is included in audioLibs?

**Reply by pschatzmann** [reply]
just re-read the link that you have added above!

**Reply by Andy2No** [reply]
> I thought I would post that here for anyone who wants to get Audio working on a esp32 feather with Platform IO!

Thanks for the example, @SnazzyBlanket.  I've been having trouble knowing how to start too.

I've been able to get this to compile for the ESP32-WROOM-DA module by commenting out two lines, which don't seem to be needed and caused errors:

```
//include "WiFi.h"
// ...
//include "AudioLibs/AudioBoardStream.h"
```

It took me a while to understand that this was using the internal DAC, outputtng in mono on GPIO25 of the ESP32-WROOM-DA module.  By changing the number of channels to two, it also came out of GPIO26 - changing this one line:

```
//AudioInfo info(44100, 1, 16);   // This gave a sine tone on GPIO25

AudioInfo info(44100, 2, 16);   // Sample Rate, channels: 2=stereo, 1=mono, bits_per_sample
  // Changing the channels to 2 gives a lower tone (one octave lower?), on both GPIO25 & GPIO26
```


FWIW, I'm using a 100k resistor per DAC channel to an RCA socket each (with the outer contact connected to ground).  I'm connecting that to a FiiO E06 headphones amp, which seems fine with potentially 3.3V peak to peak signals (based on similar experiments), and does its own AC coupling.  For more general use, it would probably be better to use a potential divider to keep it to no more than 2V peak to peak, and maybe aim for 10k output impedance.  Most preamps seem to have AC coupled inputs so there's probably no need to connect the outputs via capacitors.

**Reply by pschatzmann** [reply]
If you have an audio source that provides 1 channel, you should not just pretend that it is 2!
You would need to convert your data as [described in this chapter](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format)...

**Reply by Andy2No** [reply]
Noted.  Thanks, @pschatzmann.  Mostly, I just wanted to hear something in both ears since I'd already wired up both channels.  I may well just stick to mono though, since it needs extra processing to get stereo, or doubled up mono.

**Comment by SnazzyBlanket** [comment]
thanks so much-- re-reading the Wiki now that I have gotten things working is starting to sink in. I am going to put the Wikis on my tablet and spend the next few days really trying to understand the stream concepts a bit better. really appreciate all the help and work you have done.
